content/asp.net/getting-started/data.md

```diff
@@ -1,20 +1,168 @@
+```
+
+> Note that the `InMemory` provider is intended for testing and prototyping and is not a production data store. Each time your application starts a new copy of the database is created and, likewise, any data built up at runtime is lost when your process ends.
+
+## Collecting Some Data
+
+The next step is to start adding some data to our database, and we're going to do this through [Middleware](middleware-basic.md).  Create a folder in your project called `Middleware` and then add a new middleware class by right-clicking on the folder and selecting "Add -> New Item". Choose the Middleware class template, and name the class `ErrorCollectingMiddleware`. 
+
+Modify the constructor so that it accepts the `ErrorContext` that we've created and configured:
+
+```c#
+public ErrorCollectingMiddleware(RequestDelegate next, ErrorContext context)
+{
+    _next = next;
+    _context = context;
+}
+```
+
+Next, create a backing field for the context so that we can access it throughout our class:
+
+`private readonly ErrorContext _context;`
+
+Finally, update the `Invoke` method to the following:
+
+```c#
+public async Task Invoke(HttpContext httpContext)
+{
+    await _next(httpContext);
+
+    if (httpContext.Response.StatusCode >= 400)
+    {
+        _context.Errors.Add(new ErrorInformation
+        {
+            Path = httpContext.Request.Path,
+            StatusCode = httpContext.Response.StatusCode
+        });
+        await _context.SaveChangesAsync();
+    }
+}
+```
+
+We won't worry too much about the mechanics here for brevity, but there are two important things to note about our data and our context after we wait for the call `_next` to complete. First, whenever we see an HTTP status code at or above 400 we create a new `ErrorInformation` object on the fly and add it to the `DbSet<ErrorInformation>` property that we have in our context. Secondly, we make a call to `SaveChangesAsync` to commit our rows to the table.
+
+We have to wire in our middleware, which we've [seen before](middleware-basic.md). We want to wrap our execution around the MVC pipeline, so we can add it to the `Configure` method in our `Startup` class, just before the call to `app.UseMvc()`. We can use the extension method that was created for us in our templated middleware class:
 
-Give the reader some additional exercises/tasks they can perform to try out what they've just learned.
\ No newline at end of file
+`app.UseErrorCollectingMiddleware();`
+
+All that's left to do is for you to cause some errors! The easiest way to do this is to try to access pages that you know don't exist in your application, which gives an HTTP 404 client side error. You could also add some faulty code or throw an exception from a controller action, which would translate into an HTTP 500 server side error. As you navigate and cause problems for your application, you'll be storing the error information in your in-memory database.
```

Comment:
Yeah, I'll rephrase that. 400s are client errors, but we don't typically think of them as client errors.